src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs

```diff
@@ -146,5 +146,15 @@ private IAuthenticator CreateAuthenticator(IConnection connection, ConnectionDes
                     : new MongoDBCRAuthenticator(_credential);
 #pragma warning restore 618
         }
+
+        private IAuthenticator GetOrCreateAuthenticator(IConnection connection, ConnectionDescription description)
+        {
+            /* It is possible to have for IsMaster["SpeculativeAuthenticate"] != null and for
+             * _speculativeScramSha256Authenticator to be null in the case of multiple authenticators */
```

Comment:
I don't think this is possible. If we `_speculativeAuthenticator` is null then we didn't sent the speculative authenticate fields and we should not expect a speculative authenticate result.

Reply:
Discussed offline: possible in the case of a list of `DefaultAuthenticators` because of the way "isMaster" is merged in `DefaultAuthenticator.Authenticate`